File: packages/python/src/opensms/resources/analytics.py
"""The ``analytics`` resource."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Dict, List, Optional, Union

from ..models import AnalyticsMetrics
from ._base import Resource

DateLike = Union[str, date, datetime]


def _query(
    currency: Optional[str],
    range: Optional[str],
    from_: Optional[DateLike],
    to: Optional[DateLike],
    bucket: Optional[str],
) -> Dict[str, Any]:
    return {"currency": currency, "range": range, "from": from_, "to": to, "bucket": bucket}


class Analytics(Resource):
    """Delivery and spend metrics. Accessed as ``client.analytics``.

    Every method takes ``currency``, ``range`` (``"7d"``, 1..366 days,
    default 30d) **or** ``from_``/``to`` (datetime, date or string), and
    ``bucket`` (``day`` or ``hour``).
    """

    def overview(
        self,
        *,
        currency: Optional[str] = None,
        range: Optional[str] = None,
        from_: Optional[DateLike] = None,
        to: Optional[DateLike] = None,
        bucket: Optional[str] = None,
    ) -> AnalyticsMetrics:
        """Totals for the period."""
        q = _query(currency, range, from_, to, bucket)
        return self._t.request("GET", "/v1/analytics/overview", query=q)

    def by_country(
        self,
        *,
        currency: Optional[str] = None,
        range: Optional[str] = None,
        from_: Optional[DateLike] = None,
        to: Optional[DateLike] = None,
        bucket: Optional[str] = None,
    ) -> List[AnalyticsMetrics]:
        """Metrics per destination country."""
        q = _query(currency, range, from_, to, bucket)
        return self._t.request("GET", "/v1/analytics/by-country", query=q)

    def by_carrier(
        self,
        *,
        currency: Optional[str] = None,
        range: Optional[str] = None,
        from_: Optional[DateLike] = None,
        to: Optional[DateLike] = None,
        bucket: Optional[str] = None,
    ) -> List[AnalyticsMetrics]:
        """Metrics per carrier."""
        q = _query(currency, range, from_, to, bucket)
        return self._t.request("GET", "/v1/analytics/by-carrier", query=q)

    def by_sender_id(
        self,
        *,
        currency: Optional[str] = None,
        range: Optional[str] = None,
        from_: Optional[DateLike] = None,
        to: Optional[DateLike] = None,
        bucket: Optional[str] = None,
    ) -> List[AnalyticsMetrics]:
        """Metrics per sender ID."""
        q = _query(currency, range, from_, to, bucket)
        return self._t.request("GET", "/v1/analytics/by-sender-id", query=q)

    def timeseries(
        self,
        *,
        currency: Optional[str] = None,
        range: Optional[str] = None,
        from_: Optional[DateLike] = None,
        to: Optional[DateLike] = None,
        bucket: Optional[str] = None,
    ) -> List[AnalyticsMetrics]:
        """Metrics per time bucket."""
        q = _query(currency, range, from_, to, bucket)
        return self._t.request("GET", "/v1/analytics/timeseries", query=q)

File: packages/ruby/lib/opensms/resources/analytics.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +analytics+ resource: delivery and spend metrics. Every method takes
    # the same optional query: :currency, :range ("30d"), or :from / :to
    # (Time, Date or string; not combined with :range), and :bucket
    # (day|hour). Accessed as +client.analytics+.
    class Analytics < Base
      QUERY = %i[currency range from to bucket].freeze

      # GET /v1/analytics/overview -> Metrics + { from:, to:, currency:, environment: }.
      def overview(params = nil, **kwargs)
        fetch("overview", params, kwargs)
      end

      # GET /v1/analytics/by-country -> Array<Metrics + { key:, name: }>.
      def by_country(params = nil, **kwargs)
        fetch("by-country", params, kwargs)
      end

      # GET /v1/analytics/by-carrier -> Array<Metrics + { key:, name: }>.
      def by_carrier(params = nil, **kwargs)
        fetch("by-carrier", params, kwargs)
      end

      # GET /v1/analytics/by-sender-id -> Array<Metrics + { key:, name: }>.
      def by_sender_id(params = nil, **kwargs)
        fetch("by-sender-id", params, kwargs)
      end

      # GET /v1/analytics/timeseries -> Array<Metrics + { bucket: }>.
      def timeseries(params = nil, **kwargs)
        fetch("timeseries", params, kwargs)
      end

      private

      def fetch(name, params, kwargs)
        http_get("/v1/analytics/#{name}", Models.build(merge(params, kwargs), QUERY, times: %i[from to]))
      end
    end
  end
end
